### ISSUE TITLE
limitTo pipe on a list throws change detection exception in Dartium "checked mode" 

### ISSUE THREAD

**Issue by siggisim** [angular/angular#3321]
foo.dart
```
import 'package:angular2/angular2.dart';

@Component(
    selector: 'foo-component'
)
@View(
    templateUrl: 'package:foo.html'
)
class FooComponent {
  List fooList = ["foo", "bar", "baz"];
}
```

foo.html
```
<div *ng-for="var f of fooList | limitTo:2">
  {{f}}
</div>
```

Stack trace
```
Exception: Uncaught Error: Expression 'fooList | limitTo:2 in FooComponent: <div *ng-for="var f of fooList | limitTo:2">' has changed after it was checked. Previous value: '[foo, bar]'. Current value: '[foo, bar]' in [fooList | limitTo:2 in FooComponent: <div *ng-for="var f of fooList | limitTo:2">]
Stack Trace:
package:angular2/src/change_detection/abstract_change_detector.dart 82:5             AbstractChangeDetector.throwError
package:foo.ng_deps.dart 383:12  _FooComponent_ChangeDetector0.detectChangesInRecords
package:angular2/src/change_detection/abstract_change_detector.dart 48:10            AbstractChangeDetector._detectChanges
package:angular2/src/change_detection/abstract_change_detector.dart 68:12            AbstractChangeDetector._detectChangesInShadowDomChildren
package:angular2/src/change_detection/abstract_change_detector.dart 51:10            AbstractChangeDetector._detectChanges
package:angular2/src/change_detection/abstract_change_detector.dart 62:12            AbstractChangeDetector._detectChangesInLightDomChildren
package:angular2/src/change_detection/abstract_change_detector.dart 49:10            AbstractChangeDetector._detectChanges
package:angular2/src/change_detection/abstract_change_detector.dart 68:12            AbstractChangeDetector._detectChangesInShadowDomChildren
package:angular2/src/change_detection/abstract_change_detector.dart 51:10            AbstractChangeDetector._detectChanges
package:angular2/src/change_detection/abstract_change_detector.dart 68:12            AbstractChangeDetector._detectChangesInShadowDomChildren
package:angular2/src/change_detection/abstract_change_detector.dart 51:10            AbstractChangeDetector._detectChanges
package:angular2/src/change_detection/abstract_change_detector.dart 44:10            AbstractChangeDetector.checkNoChanges
package:angular2/src/core/life_cycle/life_cycle.dart 76:30                           LifeCycle.tick
package:angular2/src/core/life_cycle/life_cycle.dart 51:40                           LifeCycle.registerWith.<fn>
dart:async/zone.dart 904                                                             _rootRun
dart:async/zone.dart 503                                                             _ZoneDelegate.run
package:angular2/src/core/zone/ng_zone.dart 176:20                                   NgZone._run
package:angular2/src/core/zone/ng_zone.dart 191:7                                    NgZone._runUnary
dart:async/zone.dart 811                                                             _CustomZone.runUnary
dart:async/zone.dart 717                                                             _CustomZone.runUnaryGuarded
dart:async/zone.dart 743                                                             _CustomZone.bindUnaryCallback.<fn>
undefined:1undefined
```

This happens when you run dartium in checked mode, but seems to work fine with checked mode turned off:
```
DART_FLAGS='--enable-type-checks --enable-asserts' dartium
```

**Comment by TedSander** [comment]
Internal tracking bug http://b/22772734

**Comment by vsavkin** [comment]
The issue here is that limitTo creates a new collection every single time. We need to change it to create a new collection if and only if it is actually different.

**Comment by johnpapa** [comment]
any update on this?

**Comment by ericmartinezr** [comment]
@johnpapa AFAIK `limitTo` was replaced with `slice` https://github.com/angular/angular/pull/3907

**Comment by johnpapa** [comment]
thanks for the heads up.

if `limitTo` is indeed being replaced, it should be removed and this closed. 

Looking forward to ng team's take on this

**Comment by itslenny** [comment]
This might be resolved, but I'm not sure when / how.

I created the slice filter, but I wasn't aware of this issue and didn't make any effort to resolve it.

In a quick attempt to reproduce this I modified the routing test. I updated line 5 of `angular/dist/dart/examples/web/src/routing` to use a slice that returns the whole array:
```
<li *ng-for="#item of items | slice : 0" class="inbox-item-record">
```

It seems to work fine. It did not produce any errors. I encourage the ng team to confirm this, but I think this issue can be closed.
